xla/mlir_hlo/mhlo/IR/hlo_ops.cc

```diff
@@ -2013,18 +2013,27 @@ LogicalResult AllGatherOp::verify() {
   if (auto channelHandleAttr = getChannelHandleAttr())
     channelId = channelHandleAttr.getHandle();
 
-  return hlo::verifyAllGatherOp(getLoc(), getOperand(), getAllGatherDim(),
-                                getReplicaGroups(), channelId,
-                                getUseGlobalDeviceIds(), getResult());
+  if (getOperands().empty())
+    return emitOptionalError(getLoc(),
+                             "AllGather must have have at least one operand");
```

Comment:
Do we verify anywhere that there are same number of operand and results?

I'm not sure if there's an upstream MLIR trait for this, if not then let's add another check here

Reply:
Good catch! Added validation for that case to `AllGatherOp::verify`
```
  if (getNumOperands() != getNumResults())
    return emitOptionalError(
        getLoc(),
        "AllGather requires the same number of operands and results");
```

Reply:
I am surprised there isn't an MLIR core trait to check this. If not, may be adding one for MLIR might be useful (not as a part of this PR though).

Reply:
The list of OpTraits is here in LLVM [mlir/include/mlir/IR/OpDefinition.h](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/IR/OpDefinition.h#L314). I do not see a trait which compares Number of Operands and Results. 
BTW, `getOperands().empty()` check might be redundant because `SameOperandsAndResultElementType::verifyTrait` calls `OpTrait::impl::verifySameOperandsAndResultElementType(Operation *op)` which has the following
```
  if (failed(verifyAtLeastNOperands(op, 1)) ||
      failed(verifyAtLeastNResults(op, 1)))
    return failure();
```